### PR TITLE
ng_ipv6: add loopback support

### PR DESCRIPTION
**Pull request by authmillenon**
Test with your favorite `ng_netif` application with `USEMODULE="ng_ipv6_default ng_icmpv6_echo"` and `ping6` to `::1` and/or an address assigned to the interface.

### CODE REVIEW

**Comment by authmillenon** [comment]
Ping? (or maybe @OlegHahm wants to review, there are no unittests included :D)

**Comment by x3ro** [comment]
I'm so sorry I haven't gotten much RIOT related stuff done lately :< @OlegHahm could you take over this one? I will try to make some time for my other PRs soon :)

**Comment by OlegHahm** [comment]
Ok, but I cannot promise anything.

**Comment by authmillenon** [comment]
Rebased to current master.

File: sys/net/network_layer/ng_ipv6/ng_ipv6.c [hunk 1]
```diff
@@ -477,11 +487,51 @@ static void _send(ng_pktsnip_t *pkt, bool prep_hdr)
...
+    payload = ipv6->next;
 
     if (ng_ipv6_addr_is_multicast(&hdr->dst)) {
         _send_multicast(iface, pkt, ipv6, payload, prep_hdr);
     }
+    else if ((ng_ipv6_addr_is_loopback(&hdr->dst)) ||
+             ((iface == KERNEL_PID_UNDEF) && /* or registered to iface */
+              ((iface = ng_ipv6_netif_find_by_addr(&tmp, &hdr->dst)) != KERNEL_PID_UNDEF)) ||
+             ((iface != KERNEL_PID_UNDEF) &&
+              (ng_ipv6_netif_find_addr(iface, &hdr->dst) != NULL))) {
```
Comment (OlegHahm) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: please documentation to this check

File: sys/net/network_layer/ng_ipv6/ng_ipv6.c [hunk 2]
```diff
@@ -477,11 +487,51 @@ static void _send(ng_pktsnip_t *pkt, bool prep_hdr)
...
+        /* only makes sense with given interface */
```
Comment (OlegHahm) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: Unclear to what you refer here.

File: sys/net/network_layer/ng_ipv6/ng_ipv6.c (hunk 1 above)
Comment (authmillenon) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: What's unclear? The first check is documented through the function name, the second is already commented.

File: sys/net/network_layer/ng_ipv6/ng_ipv6.c (hunk 2 above)
Comment (authmillenon) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: True… even I don't know what I meant by that … will remove.

File: sys/net/network_layer/ng_ipv6/ng_ipv6.c (hunk 1 above)
Comment (OlegHahm) [inline sys/net/network_layer/ng_ipv6/ng_ipv6.c]: Aren't their three conditions? And what or who is registered to which interface?

**Comment by OlegHahm** [comment]
Just checking: loopback is not part of link local all nodes multicast address?

**Comment by authmillenon** [comment]
I'm confused about this question… Can you rephrase it maybe?

**Comment by OlegHahm** [comment]
Sorry. Basically I want to know if the node itself should respond to pings that go to ff02::1.

**Comment by OlegHahm** [comment]
Please squash.

**Comment by authmillenon** [comment]
> Sorry. Basically I want to know if the node itself should respond to pings that go to ff02::1.

Do not have any sources for this, but out of a gut feeling I'd say: No.

**Comment by authmillenon** [comment]
Squashed

**Comment by authmillenon** [comment]
Also: found some memory leak but accidentally squashed and pushed immediatly:

```diff
diff --git a/sys/net/network_layer/ng_ipv6/ng_ipv6.c b/sys/net/network_layer/ng_ipv6/ng_ipv6.c
index f54efe6..f8d40e0 100644
--- a/sys/net/network_layer/ng_ipv6/ng_ipv6.c
+++ b/sys/net/network_layer/ng_ipv6/ng_ipv6.c
@@ -498,15 +498,7 @@ static void _send(ng_pktsnip_t *pkt, bool prep_hdr)
              ((iface != KERNEL_PID_UNDEF) && /* or dst registered to given interface */
               (ng_ipv6_netif_find_addr(iface, &hdr->dst) != NULL))) {
         uint8_t *rcv_data;
-        ng_pktsnip_t *ptr = ipv6;
-        ng_pktsnip_t *rcv_pkt = ng_pktbuf_add(NULL, NULL, ng_pkt_len(ipv6),
-                                              NG_NETTYPE_IPV6);
-
-        if (rcv_pkt == NULL) {
-            DEBUG("ipv6: error on generating loopback packet\n");
-            ng_pktbuf_release(pkt);
-            return;
-        }
+        ng_pktsnip_t *ptr = ipv6, *rcv_pkt;
 
         if (prep_hdr) {
             if (_fill_ipv6_hdr(iface, ipv6, payload) < 0) {
Stage this hunk [y,n,q,a,d,/,j,J,g,e,?]? y
@@ -516,6 +508,14 @@ static void _send(ng_pktsnip_t *pkt, bool prep_hdr)
             }
         }
 
+        rcv_pkt = ng_pktbuf_add(NULL, NULL, ng_pkt_len(ipv6), NG_NETTYPE_IPV6);
+
+        if (rcv_pkt == NULL) {
+            DEBUG("ipv6: error on generating loopback packet\n");
+            ng_pktbuf_release(pkt);
+            return;
+        }
+
         rcv_data = rcv_pkt->data;
 
         /* "reverse" packet (by making it one snip as if received from NIC) */
```

**Comment by OlegHahm** [comment]
> >    Sorry. Basically I want to know if the node itself should respond to pings that go to ff02::1.

> Do not have any sources for this, but out of a gut feeling I'd say: No.

That was @emmanuelsearch's gut feeling, too and I couldn't find a source for that, either and Linux is behaving like this (no responding), too, so I guess we should be fine here.

**Comment by OlegHahm** [comment]
ACK

**Comment by OlegHahm** [comment]
What's up with Strider?

**Comment by authmillenon** [comment]
It says

```
Pull request needs squashing according to its labels set on GitHub
```

Seems like it started building faster than the labels were set ;-)

**Comment by authmillenon** [comment]
(you need to go to https://ci.riot-os.org/riot-os/riot/job/55a4f395403c20e70e1e8d77, there is a bug with the current configuration with strider, that a click on "Details" will not bring you to the related build, but tho the most current build. @phiros was already informed offline)

**Comment by OlegHahm** [comment]
I would say, let's merge it. (Cannot access strider from here anyway).

**Comment by authmillenon** [comment]
That's the current work order anyways… strider is just there to test itself ;-)

**Comment by OlegHahm** [comment]
Yes, I know. But if it's running we could also try to benefit from it. ;)
